packages/back-end/src/services/safeRolloutTimeSeries.ts

```diff
@@ -0,0 +1,210 @@
+import md5 from "md5";
+import { isFactMetricId, expandMetricGroups } from "shared/experiments";
+import { ReqContext } from "back-end/types/organization";
+import {
+  FactMetricInterface,
+  FactTableInterface,
+} from "back-end/types/fact-table";
+import { getFactTableMap } from "back-end/src/models/FactTableModel";
+import { SafeRolloutInterface } from "back-end/src/validators/safe-rollout";
+import {
+  SafeRolloutSnapshotInterface,
+  SafeRolloutSnapshotSettings,
+  MetricForSafeRolloutSnapshot,
+  SafeRolloutSnapshotAnalysisSettings,
+  SafeRolloutSnapshotMetricInterface,
+} from "back-end/src/validators/safe-rollout-snapshot";
+import {
+  CreateMetricTimeSeriesSingleDataPoint,
+  MetricTimeSeriesValue,
+  MetricTimeSeriesVariation,
+} from "back-end/src/validators/metric-time-series";
+import { logger } from "back-end/src/util/logger";
+
+export async function updateSafeRolloutTimeSeries({
+  context,
+  safeRollout,
+  safeRolloutSnapshot,
+}: {
+  context: ReqContext;
+  safeRollout: SafeRolloutInterface;
+  safeRolloutSnapshot: SafeRolloutSnapshotInterface;
+}) {
+  // TODO: Confirm if we want the same behavior here initially
+  if (
+    // Dimensioned safe rollouts are not supported
+    (safeRolloutSnapshot.dimension !== "" &&
+      safeRolloutSnapshot.dimension !== undefined) ||
+    // And no way of generating a data point if there are no metrics monitored, but shouldn't happen
+    safeRollout.guardrailMetricIds.length === 0
+  ) {
+    return;
+  }
+
+  const metricGroups = await context.models.metricGroups.getAll();
+  const metricsIds = expandMetricGroups(
+    safeRollout.guardrailMetricIds,
+    metricGroups
+  );
+
+  // TODO: Do we have more than one analysis type for safeRollout?
```

Comment:
No, we shouldn't. In the future maybe we will, but for now it's just absolute analysis. Seems like this code could be agnostic to this, though? Not sure, happy to just look at the first for now.